### ISSUE TITLE
GAT layer: attention_query_fn & attention_logit_fn:

### ISSUE THREAD

**Issue by Kristof-Neys** [deepmind/jraph#11]
hi team, I'm trying to recreate the GAT on the Cora dataset as per the paper. Am I correct to assume I will need to create the attention_query_fn and attention_logit_fn? Or would you have them defined already somewhere? Any help/example will be very much appreciated!!

**Comment by jg8610** [comment]
Hi Kristof,

Thanks for asking! We've had a few questions about the GAT so we will add an example soon to help with this.

With our current implementation, yes, you would have to define an attention_query_fn and an attention_logit_fn.

According to the paper:
https://arxiv.org/pdf/1710.10903.pdf

The attention query function is a linear layer (equation 3), and the attention is also a linear layer on the concatenated arguments with a leaky relu applied.

So I think you would have something like this (if using haiku)

```
def attention_query_fn(node_features):
  return hk.linear(embedding_size)(node_features)

def attention_logit_fn(senders, receivers, edges):
  del edges (not used in original GAT paper)
  feat = jnp.concatenate((senders, receviers), axis=-1)
  return jax.nn.leaky_relu(hk.Linear(1)(feat))
```

The paper additionally uses multi-head attention, in which case you could add a 'head' axis to the above. Let me know if that is usefl!

**Comment by Kristof-Neys** [comment]
hi Jonathan, this is very useful - thank you!!
